### PR TITLE
Graph node syncHierarchy performance

### PR DESCRIPTION
**Pull request by e-strokov**
Sync Queue stores sorted list of graph nodes to run _sync().
Dirtify methods queue node according to it hierarchy depth.
Changing hierarchy (e.g. insert child) cause changes in update order, so it is necessary to remove node from the list and queue with a new depth. There are few special cases for UI components because of that.
For a typical game scene (1-2k objects) and 10-20% objects performing _sync() it's about 1-2ms faster per frame on low end Android and IOS devices.

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

**Comment by vkalpias** [comment]
We have another class that looks quite a lot like the sync queue. It's called pc.SortedLoopArray. It basically uses binary search to insert items in a sorted order in an internal array and it also allows you to loop that array while they array is modified. The second part is probably the not needed by sync hierarchy but the sorting bit looks pretty similar... Just wandering whether we should use that since it's already used by various components and it's well tested.

File: src/scene/sync-queue.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    SyncQueue.prototype.push = function (p, v) {
+        var bs = function (index, s, e, k) {
+            if (s === e) return s;
+            var m = Math.floor((s + e) / 2);
+            if (index[m] > k)
+                return bs(index, s, m, k);
+            else if (index[m] < k)
+                return bs(index, m + 1, e, k);
+            return m;
+        };
```
Comment (willeastcott) [inline src/scene/sync-queue.js]: I'm wondering whether this function has to be allocated on every call of push. Presumably it does? Would it make sense to declare it outside the function?

File: src/scene/graph-node.js [hunk 2]
```diff
@@ -913,16 +917,32 @@ Object.assign(pc, function () {
...
+                    this._children[i]._dirtifyWorldImpl();
             }
             this._dirtyNormal = true;
             this._aabbVer++;
         },
 
+        // Sync Queue is currently in App and there is no reference to App obj in graph-node
+        // It's temporary solution to access it from global namespace
+        _queueSync: function () {
+            pc.Application.getApplication().syncQueue.push(this._graphDepth, this);
```
Comment (willeastcott) [inline src/scene/graph-node.js]: I'm not super-keen about grabbing the application like this. What do you think, @daredevildave? I do also kinda feel like the sync queue should be on the pc.Scene, not the application.

File: src/scene/sync-queue.js (hunk 1 above)
Comment (e-strokov) [reply src/scene/sync-queue.js]: Not sure if compiler can optimise it, so I'm extracting it.

File: src/scene/sync-queue.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
+Object.assign(pc, function () {
+    var SyncQueue = function () {
+        this._index = [];
+        this._values = [];
+    };
+
+    SyncQueue.prototype.runSync = function () {
+        for (var i = 0; i < this._values.length; i++) {
```
[the remark this answers is not in the document]
Comment (willeastcott) [reply src/scene/sync-queue.js]: Actually, I've done a bit more research, and the optimization I suggest about doesn't really make a measurable improvement in modern JS engines.

**Pull request merged**